ddtrace/sampler.py

```diff
@@ -348,25 +343,30 @@ def sample(self, span):
         """
         # Go through all rules and grab the first one that matched
         # DEV: This means rules should be ordered by the user from most specific to least specific
+        has_configured_rate_limit = self.limiter.rate_limit != DEFAULT_SAMPLING_RATE_LIMIT
+
+        rule_matched = False
         for rule in self.rules:
             if rule.matches(span):
                 sampler = rule
+                rule_matched = True
                 break
-        else:
-            # No rules match so use agent based sampling
-            return super(DatadogSampler, self).sample(span)
+
+        if not rule_matched:
+            sampler = RateByServiceSampler.from_datadog_sampler(self)
 
         sampled = sampler.sample(span)
-        self._set_sampler_decision(span, sampler, sampled)
 
-        if sampled:
-            # Ensure all allowed traces adhere to the global rate limit
-            allowed = self.limiter.is_allowed(span.start_ns)
-            if not allowed:
-                self._set_sampler_decision(span, self.limiter, allowed)
-                return False
+        if rule_matched or has_configured_rate_limit:
+            self._set_sampler_decision(span, sampler, sampled)
 
-        return sampled
+        # Ensure all allowed traces adhere to the global rate limit
+        allowed = self.limiter.is_allowed(span.start_ns)
```

Comment:
I think we can keep this behind the `if sampled:` gate.

The limiter is meant to limit the number of sampled traces, non-sampled traces should not be taken into account.

e.g. I want a 50% sample rate, but please don't go over 20 traces/sec sampled.

iiuc these changes change that behavior?

Reply:
I might be wrong here... seems like this is the PR which made that change

https://github.com/DataDog/dd-trace-py/commit/8c73f10f8d831ec578e0eca36bf067df1ea12f7c

I cannot figure out the motivation from the PR description, cc @majorgreys do you remember?